### ISSUE TITLE
#434 breaks on threads where `run` is not a method

### ISSUE THREAD

**Issue by bfontaine** [getsentry/sentry-python#487]
The patching code added in #434 breaks if the `run` attribute on a thread is not a method.

I’m using `sentry-sdk==0.11.1` with Python 3.7.4.

Put the following code in `demo.py`:
```python3
import sentry_sdk
from sentry_sdk.integrations.threading import ThreadingIntegration

sentry_sdk.init(default_integrations=False, integrations=[ThreadingIntegration()])

from threading import Thread

class MyThread(Thread):
    def __init__(self):
        super().__init__()
        self.run = lambda: print("hello")

t = MyThread()
t.start()
t.join()
```

Run `python3 demo.py`:
```
Traceback (most recent call last):
  File "demo.py", line 18, in <module>
    t.start()
  File "/myapp/venv/lib/python3.7/site-packages/sentry_sdk/integrations/threading.py", line 41, in sentry_start
    self.run = _wrap_run(hub_, self.run.__func__)
AttributeError: 'function' object has no attribute '__func__'
```

Remove Sentry:
```python3
from threading import Thread

class MyThread(Thread):
    def __init__(self):
        super().__init__()
        self.run = lambda: print("hello")

t = MyThread()
t.start()
t.join()
```
Run it again: `python3 demo.py`
```
hello
```

It runs fine.

----
The issue is that the patching code relies on `self.run.__func__` but [that attribute is only defined on **methods**, not on **functions**](https://docs.python.org/3/library/inspect.html#types-and-members).

You can see that with this code:
```pycon
>>> class A():
...     def f(self):
...             pass
...     def set_f(self):
...             self.f = lambda _: 42
...
>>> a = A()
>>> a.f.__func__
<function A.f at 0x106dfaef0>
>>> a.set_f()
>>> a.f.__func__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'function' object has no attribute '__func__'
```

According to [the Python docs](https://docs.python.org/3/library/inspect.html#inspect.ismethoddescriptor), one can call `inspect.ismethod` on an object to check if it’s a method and so if it has a `___func__` attribute (you could even use `hasattr`).

I haven’t tested it, but I believe something like this would do the job:
```python3
run_function = self.run.__func__ if inspect.ismethod(self.run) else self.run
self.run = _wrap_run(hub_, run_function)
```

Note that this `self.run` assigment code (already in the patch) transforms `self.run` in a function rather than a method if it were one before, meaning that in my example code `MyThread().run.__func__` is defined only if I don’t use Sentry (or at least its `Thread` integration).

-----

What do you think? I can move forward and make a pull-request with a test and a fix if the solution above suits you.

**Comment by untitaker** [comment]
Hi, thanks for bringing this to our attention. I think I am fine with a patch like that. I would still argue that setting `run` to not be anything else but a method is not "valid usage" as per the Python docs (since they say run is a method).

Even though it happens to only have to be a callable in practice. But what if there were a third-party implementation of Python that relied on `run` being a method? That impl wouldn't violate the language spec but would still not work with your code.

**Comment by bfontaine** [comment]
This patch does exactly that what you describe: it overrides `run` with a function, even if it were a method before. The fix I’m suggesting doesn’t change that, but at least it works if something else patched that method before Sentry.

**Comment by bfontaine** [comment]
Here is a piece of code that runs fine as-is, but breaks on an `AttributeError` if you use Sentry. That being said, I don’t know how to fix this issue without doing horrible Python sorcery.

```python3
from threading import Thread

class MyThread(Thread):
    def run(self):
        print(self.run.__func__)

t = MyThread()
t.start()
t.join()
```

**Comment by untitaker** [comment]
@fl4p could you check whether #516 works for you? I believe it should at least stop breaking your code, but I wonder if you see any errors even if you run with `init(debug=True)`

**Comment by untitaker** [comment]
A workaround for this is included in 0.12.3
